instrumentation/mongo/src/main/java/brave/mongo/TraceMongoCommandListener.java

```diff
@@ -0,0 +1,163 @@
+import java.net.InetSocketAddress;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A Mongo command listener that will report via Brave how long each command takes.
+ *
+ * <p>To use it, call <code>.addCommandListener(TraceMongoCommandListener.builder().tracer(tracer).build())</code>
+ * on the {@link com.mongodb.MongoClientOptions} or {@link com.mongodb.MongoClientSettings} object that is used to
+ * create the {@code MongoClient} to be instrumented.
+ */
+public class TraceMongoCommandListener implements CommandListener {
+  private final Tracer tracer;
+  private final int maxAbbreviatedCommandLength;
+  private final JsonWriterSettings jsonWriterSettings;
+  private final Map<Integer, Span> activeSpans = new ConcurrentHashMap<>();
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  protected TraceMongoCommandListener(final Builder builder) {
+    tracer = Objects.requireNonNull(builder.tracer, "tracer cannot be null");
+    maxAbbreviatedCommandLength = builder.maxAbbreviatedCommandLength;
+    jsonWriterSettings = JsonWriterSettings.builder()
+      .maxLength(Math.max(0, builder.maxAbbreviatedCommandLength))
+      .build();
+  }
+
+  @Override public void commandStarted(final CommandStartedEvent event) {
+    final Span span = tracer.nextSpan();
+    if (span.isNoop()) {
+      return;
+    }
+    activeSpans.put(event.getRequestId(), span);
+
+    final String commandName = event.getCommandName();
+    final String databaseName = event.getDatabaseName();
+    final BsonDocument command = event.getCommand();
+    final Optional<String> collectionName = getCollectionName(command);
+
+    span.name(getSpanName(commandName, collectionName))
+      .kind(Span.Kind.CLIENT)
+      .remoteServiceName(databaseName)
+      .tag("db.type", "mongo")
+      .tag("mongo.database", databaseName)
+      .tag("mongo.operation", commandName);
+
+    getAbbreviatedCommand(command).ifPresent(abbreviatedCommand -> span.tag("mongo.command", abbreviatedCommand));
+
+    collectionName.ifPresent(collection -> span.tag("mongo.collection", collection));
+
+    final InetSocketAddress socketAddress = event.getConnectionDescription().getServerAddress().getSocketAddress();
+    span.remoteIpAndPort(socketAddress.getAddress().getHostAddress(), socketAddress.getPort());
+
+    span.start();
+  }
+
+  @Override public void commandSucceeded(final CommandSucceededEvent event) {
+    final Span span = activeSpans.remove(event.getRequestId());
+    if (span == null) {
+      return;
+    }
+    span.finish();
+  }
+
+  @Override public void commandFailed(final CommandFailedEvent event) {
+    final Span span = activeSpans.remove(event.getRequestId());
+    if (span == null) {
+      return;
+    }
+    span.error(event.getThrowable());
+    span.finish();
+  }
+
+  static Optional<String> getCollectionName(final BsonDocument command) {
```

Comment:
Using the first value to get the collection name seems to rely on implementation details. The vast majority of commands seem to have a field with `commandName: collection`. I found one, `getMore`, which has a field `collection: collection`. How about trying both?

```java
BsonValue value = command.get(commandName);
if (value != null && value.isString()) {
  return value.asString().getValue().trim();
}
value = command.get("collection");
if (value != null && value.isString()) {
  // Actually we should add a static helper that coorces BsonValue into a String or null
}
```

Reply:
Actually, I also found `addShard`, and presumably similar commands, which don't operate on collections, and can have values with quite crazy cardinality as it's a comma-separated list `<replica_set>/<hostname><:port>,<hostname><:port>, ...`. In either the current or my proposed scheme, I think this will end up written as the `collection` tag.

Our builder probably should accept a whitelist of commands to have the collection annotated with a default conservative, widely used set, I think just the top three categories, aggregation, geospatial, query and write here:

https://docs.mongodb.com/manual/reference/command/#aggregation-commands